Trade

eginwong opened the pull request:
- WealthService: calculate liquidity to accept passed in props
- PropertyManagementService: get player properties by un/tradeable
- TradeService: prompt and trade details
#23

eginwong committed:
refactor:
- WealthService: calculate liquidity to accept passed in props
- PropertyManagementService: get player properties by un/tradeable
- TradeService: prompt and trade details

refactor: skip code coverage

DivvyByZero commented on ConsoleUI.js:
```diff
@@ -167,45 +186,134 @@ const consoleUI = (function (readline) {
             p.properties.length !== 0
               ? p.properties
-                  .map(propertyText)
+                  .map((p) => propertyText(p))
```
Why did we add this extra function ?

DivvyByZero replied:
Aside, I was reading some FP for JS and one of the "glue" functions was `unary = f => a => f(a)`, which would be good for use-cases the standard `.map` calls, usually: `.map(unary(propertyText))` (though for this case it wouldn't be needed - the point would to cull `.map`s extra parameters to the passed function to prevent unwanted arguments from leaking in, mais ouais)

eginwong replied:
`propertyText` was changed to binary and was causing problems because of the implicit param.

DivvyByZero commented on entities/PropertyManagementService.js:
```diff
@@ -265,6 +265,46 @@ module.exports = class PropertyManagementService {
+      // CHECK: Do not include any monopoly with houses
+      .mapValues((properties) => ({
+        properties,
+        maxBuildingCount: Math.max.apply(
```
`Math.max(...properties.map(p => p.buildings ? p.builds : 0))`

DivvyByZero commented on entities/TradeService.js:
```diff
@@ -0,0 +1,374 @@
+  static trade(UI, gameState) {
+    const { NEW, CANCEL, OFFER, ACCEPT } = TradeService.TradeStatus;
+
+    const players = gameState.players
+      .filter((p) => !p.bankrupt)
+      .map((p) => {
+        const {
+          tradeableProps,
+          untradeableProps,
+        } = PropertyManagementService.getPlayerProperties(gameState, p);
+        return {
+          ...p,
+          tradeableProps,
+          untradeableProps,
+        };
+      });
```
This might be too much, but we could do:
```javascript
const getPlayerProperties = PropertyManagementService.getPlayerProperties.bind(PropertyManagementService, gameState);

const players = gameState.players
    .filter(p => !p.bankrupt)
    .map(p => [p, getPlayerProperties(p)])
    .map(([p, { tradeableProps, untradeableProps }]) => ({ ...p, tradeableProps, untradeableProps }));
```

eginwong committed:
feat: trade main rule logic

refactor: allow for better console message

fix: dereferenced property does not update ownedBy

fix: incorrect card trade reference

refactor: trade service bind

eginwong committed:
refactor: stylistic changes

eginwong committed:
refactor: use UI method in highest rollign player

DivvyByZero commented on ConsoleUI.js:
```diff
@@ -167,45 +186,159 @@ const consoleUI = (function (readline) {
+        cliTable.push(row);
+      });
+
+      console.log(cliTable.toString());
```
Shouldn't this be a `UI` call or is this debugging ?

DivvyByZero replied:
Nvm I am drunk (not actually, but I realized this is the UI file...)

DivvyByZero reviewed (approved):
LGTM
